airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py

```diff
@@ -0,0 +1,141 @@
+    """Callable that help to match request object with call rate policies."""
+
+    def __call__(self, request: Any) -> bool:
+        pass
+
+
+class HttpRequestMatcher(RequestMatcher):
+    """Simple implementation of RequestMatcher for http requests case"""
+
+    def __init__(self, method: str = None, url: str = None, params: Optional[dict] = None, headers: Optional[dict] = None):
+        self._method = method
+        self._url = url
+        self._params = params
+        self._headers = headers
+
+    @staticmethod
+    def _match_dict(obj: dict, pattern: dict) -> bool:
+        """Check that all elements from pattern dict present and have the same values in obj dict
+
+        :param obj:
+        :param pattern:
+        :return:
+        """
+        return pattern.items() <= obj.items()
+
+    def __call__(self, request: Any) -> bool:
+        if isinstance(request, (requests.Request, requests.PreparedRequest)):
+            if self._method is not None:
+                if request.method != self._method:
+                    return False
+            if self._url is not None:
+                if request.url != self._url:
+                    return False
+            if self._params is not None:
+                if not self._match_dict(request.params, self._params):
+                    return False
+            if self._headers is not None:
+                if not self._match_dict(request.headers, self._headers):
+                    return False
+            return True
+
+        return False
+
+
+class AbstractAPIBudget(abc.ABC):
+    """Interface to some API where client allowed to have N calls per T interval.
+
+    Important: APIBudget is not doing any API calls, the end user code is responsible to call this interface
+        to respect call rate limitation of the API.
+
+    It supports multiple policies applied to different group of requests. To distinct these groups we use RequestMatchers.
+    Individual policy represented by CallRatePolicy and currently supports only moving window strategy.
+    """
+
+    @abc.abstractmethod
+    def add_policy(self, request_matcher: RequestMatcher, policy: CallRatePolicy):
+        """Add policy for calls
+
+        :param request_matcher: callable to match request object with corresponding policy
+        :param policy: to acquire calls
+        :return:
+        """
+
+    @abc.abstractmethod
+    def acquire_call(self, request: Any, wait: Optional[int] = None) -> bool:
+        """Try to get a call from budget
+
+        :param request:
+        :param wait: if set >0 will wait number of seconds; if wait == 0 - will return immediately; if wait is None (default) - will block
+        :return: True if call was acquired, False - otherwise
+        """
+
+
+class APIBudget(AbstractAPIBudget):
+    def __init__(self):
+        self._policies: list[tuple[RequestMatcher, CallRatePolicy]] = []
+
+    def add_policy(self, request_matcher: RequestMatcher, policy: CallRatePolicy):
```

Comment:
would this support advanced use cases like dynamically setting a new rate limit (eg if we receive a response with a x-rate-limit header )?

Reply:
I was thinking about this case as part of call rate functionality. But I'm not convinced that these use cases should be solved with the same tool. I'm open a discussion here, and gladly will hear arguments.

My thoughts so far.

1. The main benefit of call_rate module is that we can set limits lower then actual API restrictions, therefore don't fully utilize API call budget.
2. Reacting to API state and call rate limit hit also falls into the backoff and retry handlers category.
3. We can initialize call buckets from x-rate-limit, i.e. set current usage, and save timestamp of limit reset, but there is a conflict between tracking connector usage and API utilization.
4. ...

Reply:
For context: I don't think we need to support that use case yet, but I expect there will be a need for dynamically setting the rate limit so I want to make sure we have a plan to support it.

An use case I have in mind is when an API might have a different rate limit depending on the user's plan.

We might start with no limit, and adjust the rate based off the response
```
  response = session._send(...)
  self._rate_limit_update_rate(response)
  yield response
```

Reply:
On the other hand, if particular implementation of CallRatePolicy will completely rely on x-rate-limit headers, then we can just use previous response and hope that other client didn't use all budget of API withing current window.
This will be very simple implementation without pyratelimiter and so on, will just keep a state of last real request non-cached and limit reset timestamp.

Reply:
@keu, so we'd just need to implement a stateful policy that updates itself when `acquire_call` is called?

that works for me 👍

Reply:
yes, correct, the particular implementation of CallRatePolicy can have modifiers and provide a new list of Rates.

I think the cleanest way would be to remove the existing CallRatePolicy and add a new one, but the problem is that the order of adding matters for request matchers - by default first request matcher will win and therefore we will need to re-insert all matches with policies after that one.

Reply:
if iterating over all the policies and matchers like we do in [acquire_call](https://github.com/airbytehq/airbyte/pull/31276/files#diff-e21ba47ba176b1cf99b7ad8c7b003847b710bde1082561515560928c4275a8f0R205-R206), won't we always block on the stricter policy that applies to the request?

Reply:
@girarda it will block on the first policy that matches the request,
I have added the explanation above, why implementing lookup for stricter requests matches is challenging